src/mlpack/methods/ann/layer/neural_turing_machine.hpp

```diff
@@ -0,0 +1,256 @@
+   * forward pass.
+   *
+   * @param input The propagated input activation.
+   * @param gy The backpropagated error.
+   * @param g The calculated gradient.
+   */
+  template<typename eT>
+  void Backward(const arma::Mat<eT>&& /* input */,
+                arma::Mat<eT>&& gy,
+                arma::Mat<eT>&& g);
+
+  /*
+   * Calculate the gradient using the output delta and the input activation.
+   *
+   * @param input The input parameter used for calculating the gradient.
+   * @param error The calculated error.
+   * @param gradient The calculated gradient.
+   */
+  template<typename eT>
+  void Gradient(arma::Mat<eT>&& input,
+                arma::Mat<eT>&& /* error */,
+                arma::Mat<eT>&& /* gradient */);
+
+  /*
+   * Resets the cell to accept a new input.
+   * This breaks the BPTT chain starts a new one.
+   */
+  void ResetCell();
+
+  //! The value of the deterministic parameter.
+  bool Deterministic() const { return deterministic; }
+  //! Modify the value of the deterministic parameter.
+  bool& Deterministic() { return deterministic; }
+
+  //! Get the parameters.
+  OutputDataType const& Parameters() const { return weights; }
+  //! Modify the parameters.
+  OutputDataType& Parameters() { return weights; }
+
+  //! Get the input parameter.
+  InputDataType const& InputParameter() const { return inputParameter; }
+  //! Modify the input parameter.
+  InputDataType& InputParameter() { return inputParameter; }
+
+  //! Get the output parameter.
+  OutputDataType const& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the gradient.
+  OutputDataType const& Gradient() const { return gradient; }
+  //! Modify the gradient.
+  OutputDataType& Gradient() { return gradient; }
+
+  //! Get the model modules.
+  std::vector<LayerTypes>& Model() { return network; }
+
+  /**
+   * Serialize the layer
+   */
+  template<typename Archive>
+  void Serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  //! Locally-stored number of input units.
+  size_t inSize;
+
+  //! Locally-stored number of output units.
+  size_t outSize;
+
+  //! Number of memory locations.
+  size_t numMem;
+
+  //! Size of memory locations.
+  size_t memSize;
+
+  //! Circular shift rotation size.
+  size_t shiftSize;
+
+  //! Memory read layer.
+  LayerTypes readMem;
+
+  //! Memory write layer.
+  LayerTypes writeMem;
+
+  //! Controller.
+  LayerTypes controller;
+
+  //! Locally stored memory content error.
+  arma::mat dMem;
+
+  //! Locally stored previous memory content error.
+  arma::mat dMemPrev;
+
+  //! All zeros vector for initializing read.
+  arma::mat allZeros;
+
+  //! All ones matrix for initializing memory.
+  arma::mat allOnes;
```

Comment:
You can ignore this comment, we discussed it in the GRU PR.  I agree that this is the best way we can handle this for now.  I forgot to delete it from the review before I hit 'post' :)